dlpy/applications/enet.py

```diff
@@ -0,0 +1,271 @@
+
+    conv1 = Conv2d(reduced_depth, 3, stride=2, padding=1, act='identity', include_bias=False)(x)
+    bn1 = BN(act='relu')(conv1)
+
+    conv2 = Conv2d(reduced_depth, 3, stride=1, act='identity', include_bias=False)(bn1)
+    bn2 = BN(act='relu')(conv2)
+
+    conv3 = Conv2d(out_depth, 1, stride=1, act='identity', include_bias=False)(bn2)
+    bn3 = BN(act='relu')(conv3)
+
+    pool1 = Pooling(2, stride=2)(x)
+    conv4 = Conv2d(out_depth, 1, stride=1, act='identity', include_bias=False)(pool1)
+    bn4 = BN(act='relu')(conv4)
+
+    res1 = Res()([bn3, bn4])
+
+    return res1
+
+
+def regularBNeck(x, in_depth, out_depth, projection_ratio=4):
+    '''
+    Defines the regular bottleneck of ENet
+
+    Parameters
+    ----------
+    x :
+       Previous layer to this block
+    in_depth : int
+       Depth of the layer fed into this block
+    out_depth : int
+       Depth of the output layer of this block
+    projection_ratio : int, optional
+       Used to calculate the reduced_depth for intermediate convolution layers
+       Default: 4
+
+    Returns
+    -------
+    :class:`Res`
+    '''
+
+    reduced_depth = int(in_depth // projection_ratio)
+
+    conv1 = Conv2d(reduced_depth, 1, stride=1, act='identity', include_bias=False)(x)
+    bn1 = BN(act='relu')(conv1)
+
+    conv2 = Conv2d(reduced_depth, 3, stride=1, act='identity', include_bias=False)(bn1)
+    bn2 = BN(act='relu')(conv2)
+
+    conv3 = Conv2d(out_depth, 1, stride=1, act='identity', include_bias=False)(bn2)
+    bn3 = BN(act='relu')(conv3)
+
+    res1 = Res()([bn3, x])
+
+    return res1
+
+
+def upsamplingBNeck(x, in_depth, out_depth, projection_ratio=4):
+    '''
+    Defines the up-sampling bottleneck of ENet
+
+    Parameters
+    ----------
+    x :
+       Previous layer to this block
+    in_depth : int
+       Depth of the layer fed into this block
+    out_depth : int
+       Depth of the output layer of this block
+    projection_ratio : int, optional
+       Used to calculate the reduced_depth for intermediate convolution layers
+       Default: 4
+
+    Returns
+    -------
+    :class:`BN`
+    '''
+
+    reduced_depth = int(in_depth // projection_ratio)
+
+    conv1 = Conv2d(reduced_depth, 1, stride=1, act='identity', include_bias=False)(x)
+    bn1 = BN(act='relu')(conv1)
+
+    tconv1 = Conv2DTranspose(reduced_depth, 3, stride=2, padding=1, output_padding=1, act='identity',
+                             include_bias=False)(bn1)
+    bn2 = BN(act='relu')(tconv1)
+
+    conv3 = Conv2d(out_depth, 1, stride=1, act='identity', include_bias=False)(bn2)
+    bn3 = BN(act='relu')(conv3)
+
+    return bn3
+
+
+def ENet(conn, model_table='ENet', n_classes=2, n_channels=3, width=512, height=512, scale=1.0/255,
+         norm_stds=None, offsets=None, random_mutation=None, init=None, random_flip=None, random_crop=None,
+         output_image_type=None, output_image_prob=False):
+    '''
+    Generates a deep learning model with the E-Net architecture.
```

Comment:
can we please add a reference here

Reply:
Reference was added to line 221 following the format of other model types.